### PR TITLE
feat: exome option and changes to vardict options

### PR DESCRIPTION
**Pull request by mathiasbio**
## Description

There are a few related changes in this PR, all connected to modifications to the VarDict output.
1. An extra option has been added to balsamic `--exome` in order to allow for different variant-calling filters for exome-samples which usually have much lower coverage than the smaller panels. 
2. bcftools filters for exome samples have been changed from standard targeted analysis, most importantly the minimum depth has been decreased from 100 to 20. https://github.com/Clinical-Genomics/BALSAMIC/issues/1411
3. stringency of bcftools filters for TGA samples have been somewhat reduced, reducing min MQ from 40 to 30.
4.  "-u" flag for VarDict has been removed as it could result in wrongful strandbias https://github.com/Clinical-Genomics/BALSAMIC/issues/1396
5. "-U" flag has been removed VarDict tumor-normal, this option disables SV calling and it has been added back due FLT3 calling which requires SV functionality of VarDict to be enabled

#### Added

* new option for exome samples --exome with modified bcftools filters compared to standard targeted workflow 

#### Changed

* reduced stringency of targeted none-exome bcftools filters for min MQ
* removed -u flag from VarDict T+N and T only rules

#### Fixed

- [Description]

#### Removed

* removed -U flag to VarDict T+N rule to start calling SVs in VarDict (required for FLT3)

## Documentation

- [ ] N/A
- [ ] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - [Document Name]

### Feature Tests

- [ ] N/A
- [ ] Test [Description]
    - [Screenshot]

#### Variant-calling quality testing

Detailed results from testing can be found in this sheet on CG google drive: https://docs.google.com/spreadsheets/d/1-CnG0L2J05Z4Phzdird6JKSXCbG_w-O_xOJfpJNTJ70/edit#gid=0

- [x] changes appear to result in an improved analysis quality, or at least not worse based on looking at true positive calls. **Comment:** See "sensitivity and precision TWISTpancancer" in above sheet, a significant improvement which likely comes predominantly from removing the "-u" flag.
- [x] changes in the number of variants called are not significant, or justified below. **Comment**: There is an increase in number of variants called, in some cases more than others, in particular for the exome cases tested, and for one normal TGA case. In the exome cases this is likely due to the lowering of the depth from 100 to 20, and in the case of the TGA case it's most likely due to the removal of the "-u" flag and the associated low insert size of that sample which probably resulted in a lot of variants being filtered out due to artificial strand-bias. 
- [x] significant changes in the number of variants called is justified because: 
- **TGA smaller panels:** with regards to the TGA samples, the quality of the analysis has been assessed using the TWISTpancancer reference samples and the sensitivity has been much improved to comparably little cost to precision. The only question-mark to me is the effect of lowering the MQ from 40 to 30, which resulted in the detection of 2 more true variants, and 38 more FPs (however, these TWIST samples do not have accurate FP detection...so they might also be true). 
- **TGA exome:** with regards to the exome samples, it seems that we are seeing a more signfiicant increase in the number of variants called based on simply adding the "-u" flag. With an about 300% increase in number of variants even without reducing the min depth to 20. After then reducing the min depth to 20 from 100, we get an about 30% increase in variants, which would make sense proportionally as these exome samples were covered at about 70% above 100X. 

##### Further notes from testing:

**Comparatively larger effect of removal "-u" for exome samples**

The number of variants increased more substantially for the exome samples run in comparison to the other TGA-samples (except 1), even without reducing the DP to 20. This is likely due to the same reason for all these samples, a smaller insert-size which may have pushed more variants to beyond the threshold for strand-bias when using VarDict's "-u" option. 

**Further info regarding lowering MQ40 to MQ30**

I looked at about half of the extra 27 variants called in one of the TGA T+N cases in IGV and the majority of them look perfectly fine, and a few looked perhaps questionable. Screenshots are available in the sheet "MQ30-40"

**Further info regarding keeping SVs in the T+N workflow**

I looked at 3 T+N cases used in this testing, and summarising across all final variant-calls in these cases 6 out of 8000 variants were SVs. 

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [ ] N/A
    - [x] Verified multiple times
- **TGA T/N Workflow**
    - [ ] N/A
    - [x] Verified multiple times
 - **TGA T/O Workflow with --exome**
    - [ ] N/A
    - [x] Verified multiple times
- **TGA T/N Workflow with --exome**
    - [ ] N/A
    - [x] Verified multiple times
- **UMI T/O Workflow**
    - [ ] N/A
    - [x] Verified multiple times
- **UMI T/N Workflow**
    - [ ] N/A
    - [x] Verified multiple times
- **WGS T/O Workflow**
    - [x] N/A (NOTHING HAS CHANGED HERE)
    - [ ] Verified
- **WGS T/N Workflow**
    - [ ] N/A  (NOTHING HAS CHANGED HERE, **VERIFIED ONCE ANYWAY**)
    - [x] Verified once
- **QC Workflow**
    - [x] N/A  (NOTHING HAS CHANGED HERE)
    - [ ] Verified
- **PON Workflow**
    - [x] N/A  (NOTHING HAS CHANGED HERE)
    - [ ] Verified

## Clinical Genomics Stockholm

<details>
<summary></summary>

### Documentation

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### User Changes

- [ ] N/A
- [x] This PR affects the output files or results.
    - [x] User feedback is considered unnecessary because [Justification below].
    - [x] Affected users have been included in the development process and given a chance to provide feedback. **Comment:** The users have not been consulted after the fix was made, but they were the ones essentially requesting this update as important variants were filtered out. With regards to the details of the filters I judge that we are better suited as the experts on the bioinformatics, and knowing better the limits of our truthsets and the different coverage-depths we have for different applications, etc...

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

</details>

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [x] Provided a comprehensive description of the PR.
    - [x] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [x] Verified and updated documentation if necessary.
- **Tests**
    - [x] Described and tested the functionality addressed in the PR.
    - [x] Ensured integration of the new code with existing workflows.
    - [x] Confirmed that meaningful unit tests were added for the changes introduced.
    - [x] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [ ] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [x] Code implements the intended features or fixes the reported issue.
    - [x] Code follows the project's coding standards and style guide.
- **Documentation**
    - [x] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Tests**
    - [x] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [x] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

File: BALSAMIC/utils/cli.py [hunk 1]
```diff
@@ -487,6 +487,12 @@ def get_analysis_fastq_files_directory(case_dir: str, fastq_path: str) -> str:
         return analysis_fastq_path.as_posix()
     return Path(fastq_path).as_posix()
 
+def validate_exome_option(ctx: click.Context, exome: bool):
```
Comment (ivadym) [inline BALSAMIC/utils/cli.py:490]: Nice 🔥

File: CHANGELOG.rst [hunk 2]
```diff
@@ -4,13 +4,17 @@
 Added:
 ^^^^^^
 * high_normal_tumor_af_frac filter in bcftools for TNscope T+N filtering out more than 30% TINC https://github.com/Clinical-Genomics/BALSAMIC/pull/1289
+* new option for exome samples --exome with modified bcftools filters compared to standard targeted workflow
+* added -U flag to VarDict T-only rule to not call SVs
```
Comment (ivadym) [inline CHANGELOG.rst:8]: Would fit better in `changes`, as you're replacing -u with -U to not call SVs

File: BALSAMIC/constants/variant_filters.py [hunk 3]
```diff
@@ -10,15 +10,15 @@
     "description": "General purpose filters used for filtering any variant caller",
 }
 
-# Configuration of VARDICT settings
-VARDICT_SETTINGS = {
+# Configuration of VARDICT settings for smaller panels
+VARDICT_SETTINGS_PANEL = {
```
Comment (ivadym) [inline BALSAMIC/constants/variant_filters.py:14]: Maybe we could avoid repetition here by reusing the common parameters:
```
VARDICT_COMMON_SETTINGS = { ... }

VARDICT_PANEL_SETTINGS = {
    **common_settings,
    ...
}

VARDICT_EXOME_SETTINGS = {
    **common_settings,
    ...
}
```

File: BALSAMIC/workflows/balsamic.smk [hunk 4]
```diff
@@ -115,7 +116,13 @@ else:
 
 # Varcaller filter settings
 COMMON_FILTERS = VarCallerFilter.model_validate(COMMON_SETTINGS)
-VARDICT = VarCallerFilter.model_validate(VARDICT_SETTINGS)
+
+# Set VarDict settings depending on if panel is exome or not
+if config_model.panel and config_model.panel.exome:
+    VARDICT = VarCallerFilter.model_validate(VARDICT_SETTINGS_EXOME)
+else:
+    VARDICT = VarCallerFilter.model_validate(VARDICT_SETTINGS_PANEL)
```
Comment (ivadym) [inline BALSAMIC/workflows/balsamic.smk:124]:
```suggestion
    VARDICT = VarCallerFilter.model_validate(VARDICT_SETTINGS_PANEL)
    
if config_model.panel.exome:
    VARDICT = VarCallerFilter.model_validate(VARDICT_SETTINGS_EXOME)

```

**Review by ivadym — COMMENTED** [review]
⚡ 🔥

File: BALSAMIC/workflows/balsamic.smk (hunk 4 above)
Comment (mathiasbio) [reply BALSAMIC/workflows/balsamic.smk:124]: Won't that crash for WGS where the panel model hasn't been created?

File: BALSAMIC/constants/variant_filters.py (hunk 3 above)
Comment (mathiasbio) [reply BALSAMIC/constants/variant_filters.py]: Done!

File: CHANGELOG.rst (hunk 2 above)
Comment (mathiasbio) [reply CHANGELOG.rst:8]: But they're two quite different changes, one is to remove the risk of wrongly filtering out variants due to strand-bias, and the other is to not call structural variants, feels like it could be confusing to put them under the same change.

Comment (mathiasbio) [reply CHANGELOG.rst:8]: but it can be quite cluttered if we basically try to document every little change we make in the changelog, so maybe it's better to have some more general summary like "changed vardict settings"

Comment (ivadym) [reply CHANGELOG.rst]: Maybe then just clarify the removal of the -u flag. I think it's nice if we're specific, it helps to write the message to the customers and to follow the pipeline development

File: BALSAMIC/workflows/balsamic.smk (hunk 4 above)
Comment (ivadym) [reply BALSAMIC/workflows/balsamic.smk:125]: Isn't this the current behaviour? We're defining `VARDICT_SETTING` but not using it. It does not rely on the panel model

Comment (mathiasbio) [reply BALSAMIC/workflows/balsamic.smk:125]: The if config_model.panel.exome part is new, and we didn't have that if-statement before we just validated the vardict settings :D But if the panel model hasn't been populated doesn't looking inside it cause some attribute error?

Comment (ivadym) [reply BALSAMIC/workflows/balsamic.smk:125]: Oh you mean the if statement? Yes it should be `if config_model.panel and config_model.panel.exome:`. I just meant to say that the default assignation should be defined outside the `if` so we avoid having an else

Comment (mathiasbio) [reply BALSAMIC/workflows/balsamic.smk:125]: Oh god, of course. That makes sense...I'll make the adjustment ❤️ Thanks!

File: docs/balsamic_filters.rst [hunk 5]
```diff
@@ -109,15 +104,21 @@ Somatic Callers for reporting SNVs/INDELS
 The results of `Vardict` variant calling are further post-filtered based on several criteria (`Vardict_filtering`_) to retrieve high-confidence variant calls.
 These high-confidence variant calls are the final list of variants uploaded to Scout or available in the delivered VCF file in Caesar.
 
+
+There are two slightly different post-processing filters used depending on if the flag "--exome" is supplied at the start, which is designed to accomodate targeted analyses of lower depths.
```
Comment (ivadym) [inline docs/balsamic_filters.rst:108]: Do we need to be so technical here (im refering to the --exome flag)? Maybe it's clearer if we refer to it as exome samples in general, what do you think?

File: BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule [hunk 6]
```diff
@@ -32,7 +32,7 @@ mkdir -p {params.tmpdir};
 export TMPDIR={params.tmpdir};
 export VAR_DICT_OPTS='\"-Djava.io.tmpdir={params.tmpdir}\" \"-Xmx90G\"';
 
-vardict-java -U -u -I 600 -G {input.fa} -f {params.af} -N {params.case_name} \
+vardict-java -I 600 -G {input.fa} -f {params.af} -N {params.case_name} \
```
Comment (ivadym) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: I'm not sure about removing the -U flag, specially if we are planning to make a release after this PR.

If we want Vardict to call SVs for T/N, we should have an additional file and fed it into SVDB. Same thing for T/O.

Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: Then we could also address this issue: https://github.com/Clinical-Genomics/BALSAMIC/issues/1344

But I'm unsure about the effort though

**Review by ivadym — COMMENTED** [review]
Impressive amount of work! Thanks for documenting everything ⭐

File: BALSAMIC/constants/variant_filters.py [hunk 7]
```diff
@@ -35,12 +27,36 @@
...
     "filter_type": "general",
-    "analysis_type": "tumor_only",
+    "analysis_type": "tumor_only,tumor_normal",
     "description": "General purpose filters used for filtering VarDict",
 }
 
+# Configuration of VARDICT settings for smaller panels
+VARDICT_SETTINGS_PANEL = {
+    **VARDICT_SETTINGS_COMMON,
+    "AD": {"tag_value": 5, "filter_name": "balsamic_low_tumor_ad", "field": "INFO"},
```
[the remark this answers is not in the document]
Comment (mathiasbio) [reply BALSAMIC/constants/variant_filters.py:41]: Not really! I could move it, in some ways I just liked the idea of having the parameters already separated to encourage creating more specific parameters for the two versions. But it's quite easy to just move them whenever we feel the need as well, so I'll just move it up :  )

File: BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule (hunk 6 above)
Comment (mathiasbio) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: I also don't think it's going to be such a big mess, since it was only quite recently that it was brought up as a bug that we saw SVs in VarDict in Scout, when it had been around for around 3 years already

File: BALSAMIC/utils/cli.py [hunk 8]
```diff
@@ -487,6 +487,12 @@ def get_analysis_fastq_files_directory(case_dir: str, fastq_path: str) -> str:
         return analysis_fastq_path.as_posix()
     return Path(fastq_path).as_posix()
 
+def validate_exome_option(ctx: click.Context, _param: click.Parameter, exome: bool):
```
[the remark this answers is not in the document]
Comment (mathiasbio) [reply BALSAMIC/utils/cli.py:491]: Done!

File: tests/utils/test_utils.py [hunk 9]
```diff
@@ -1001,6 +1002,28 @@ def test_get_fastp_parameters(balsamic_model: ConfigModel):
...
+    # THEN exome argument should be correctly set
+    assert validate_exome_option(ctx, None, True) == True
+
+    # GIVEN that a panel bedfile has NOT been supplied and exome parameter set to true
+    ctx = MockClickContext({"panel_bed": None})
+
+    # WHEN validating exome option
+    # THEN a bad parameter error should be raised
+    with pytest.raises(click.BadParameter):
+        validate_exome_option(ctx, None, True)
```
Comment (ivadym) [inline tests/utils/test_utils.py:1024]: That's a creative way of doing this but I have Sonarlint complaining here: `validate_exome_option(ctx, None, True)`

I think this an be simplified by just using click.Context and click.Patameters. Same as we doing for `test_validate_cache_version_develop` for example.

Comment (ivadym) [inline tests/utils/test_utils.py:1024]: Let's use named arguments for function with > 1 parameter

File: BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule (hunk 6 above)
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: Yes, that's true. However, because of the same issue, customers may unexpectedly find SVs in their SNVs tab in Scout for tumor-normal cases.

I definitely don't think we should address this in the current pull request. But if we plan to release after merging this one, our options could be:

- Assess the impact of having these SVs in tumor-normal cases and gather customer feedback.
- Acknowledge it as a known issue and plan to fix it later by separating SVs and SNVs or exploring alternative methods to vardict for SV detection.
- Consider including this user story for the next release: https://github.com/Clinical-Genomics/BALSAMIC/issues/1344

File: tests/utils/test_utils.py (hunk 9 above)
Comment (mathiasbio) [reply tests/utils/test_utils.py]: I think I figured it out, thanks! :D

File: BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule (hunk 6 above)
Comment (mathiasbio) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: I agree! It would definitely be a lot more cleaner if we could remove the SVs from this file entirely, we would just need to talk to the customers and make sure that they are aware. Shouldn't be that difficult! : )

Comment (mathiasbio) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: I put it in milestone for release 16

Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: Did you also check how many extra variants we get in the TGA tumor-normal cases just by keeping the SVs?

Might be worth adding that info to the test section in this PR 🤔

Comment (mathiasbio) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: I didn't specifically look for extra variants from that flag, but I think they're in the scale of 0-10 but I'll verify!

**Review by ivadym — APPROVED** [review]
Great work! 💯 🕵️ 🥇

Please update the PR with the number of SVs we're including, so we can keep track of it. Nothing more from my side 😄

File: BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule (hunk 6 above)
Comment (mathiasbio) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: Seems that about 6 variants in 8000 are extra SVs

Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule:35]: Thanks mathias!
